### PR TITLE
Enrich PAN-validation with State IIN logic

### PR DESCRIPTION
**Pull request by devinmorgan**
# Description
Please include a summary of the change. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Also
- Add any other relevant change made on this PR.

## Notes
- Describe your changes

## Tests Added / Updated?
- YES or NO. If not, why?

## Screenshots

### CODE REVIEW

**Comment by devinmorgan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #57** <a href="https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/57" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/57?utm_source=stack-comment).

**Comment by dleis612** [comment]
Also, can you get rid of the call to set `textInputLayout.error` in this PR? That was a specific request from GoPuff and it's something we never should have done in the first place.

File: forage-android/src/main/java/com/joinforage/forage/android/core/element/state/PanElementStateManager.kt [hunk 1]
```diff
@@ -1,26 +1,63 @@
...
+private fun hasInvalidStateIIN(cardNumber: String) : Boolean {
+    return queryForStateIIN(cardNumber) == null
+}
+private fun iinTooShort(cardNumber: String): Boolean {
+    val iin = queryForStateIIN(cardNumber) ?: return false
+    return cardNumber.length < iin.panLength
+}
+private fun iinTooLong(cardNumber: String): Boolean {
+    val iin = queryForStateIIN(cardNumber) ?: return false
+    return cardNumber.length > iin.panLength
```
[the remark this answers is not in the document]
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/element/state/PanElementStateManager.kt]: no issue with this! I had something like that before but probably started playing too much code golf ⛳ . Fixed in [1240bd2](https://github.com/teamforage/forage-android-sdk/pull/57/commits/1240bd2adbf50394ab6dd6a6d46f31d320f1c14c)

File: forage-android/src/main/java/com/joinforage/forage/android/core/element/state/PanElementStateManager.kt [hunk 2]
```diff
@@ -1,26 +1,63 @@
...
-            isValid = false
-            validationError = IncompleteEbtPanError
+        isValid = cardNumber.isEmpty() || passesValidation(cardNumber)
+    }
+
+    private fun setValidationError(cardNumber: String) {
+        validationError = if (cardNumber.isEmpty()) {
+            null
+        } else if (missingStateIIN(cardNumber)) {
+            InvalidEbtPanError
```
[the remark this answers is not in the document]
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/element/state/PanElementStateManager.kt]: This makes sense. It's both invalid and incomplete but incomplete is less jarring than invalid if you have fewer than 6 characters. Good call - fixed in [1240bd2](https://github.com/teamforage/forage-android-sdk/pull/57/commits/1240bd2adbf50394ab6dd6a6d46f31d320f1c14c)

**Comment by devinmorgan** [comment]
> Also, can you get rid of the call to set `textInputLayout.error` in this PR? That was a specific request from GoPuff and it's something we never should have done in the first place.

No issue with that @dleis612  - been eager to get rid of that error as well. I'll make a stacked PR on top of this to keep things clean. Thanks for flagging this!

File: forage-android/src/test/java/com/joinforage/forage/android/core/element/state/PanElementStateManagerTest.kt [hunk 3]
```diff
@@ -13,44 +15,171 @@ class PanSetIsValidTest {
...
+        assertThat(state.isValid).isFalse
+    }
+
+    @Test
+    fun `cardNumber has valid state IIN but is shorter than expected length`() {
+        val tooShortMaineNumber: String = "507703111" // Maine is 507703
+        val manager = PanElementStateManager.forEmptyInput()
+        manager.handleChangeEvent(tooShortMaineNumber)
+        val state = manager.getState()
+        assertThat(state.isValid).isFalse
```
Comment (dleis612) [inline forage-android/src/test/java/com/joinforage/forage/android/core/element/state/PanElementStateManagerTest.kt:44]: A card that is a valid EBT card, but hasn't reached the correct length yet should be valid and incomplete

File: forage-android/src/test/java/com/joinforage/forage/android/core/element/state/PanElementStateManagerTest.kt [hunk 4]
```diff
@@ -13,44 +15,171 @@ class PanSetIsValidTest {
...
+        val okMaineNumber: String = "5077031111111111111" // Maine is 507703
+        val manager = PanElementStateManager.forEmptyInput()
+        manager.handleChangeEvent(okMaineNumber)
         val state = manager.getState()
         assertThat(state.isValid).isTrue
-        assertThat(state.validationError).isNull()
+    }
+
+    @Test
+    fun `cardNumber has valid state IIN and is too long`() {
```
Comment (dleis612) [inline forage-android/src/test/java/com/joinforage/forage/android/core/element/state/PanElementStateManagerTest.kt:57]: I blocked the user from being able to type any text longer than the `min(expectedLen, 19)`

**Review by dleis612 — CHANGES_REQUESTED** [review]

File: forage-android/src/test/java/com/joinforage/forage/android/core/element/state/PanElementStateManagerTest.kt (hunk 3 above)
Comment (devinmorgan) [reply forage-android/src/test/java/com/joinforage/forage/android/core/element/state/PanElementStateManagerTest.kt:44]: you, sir, are completely correct - will address this now

Comment (devinmorgan) [reply forage-android/src/test/java/com/joinforage/forage/android/core/element/state/PanElementStateManagerTest.kt]: fixed in [568f829](https://github.com/teamforage/forage-android-sdk/pull/57/commits/568f829debb13972c3c706cdbf1e22e4b0fa2a7b)

**Review by dleis612 — APPROVED** [review]

**Pull request merged**